ufl/finiteelement/hdivcurl.py

```diff
@@ -143,6 +143,10 @@ def reference_value_shape(self):
     def mapping(self):
         return self._mapping
 
+    def sobolev_space(self):
+        """Return the underlying Sobolev space."""
+        return self.wrapee.sobolev_space()
```

Comment:
I think this one is wrong. If you change the pullback then in general you're going to change the continuity of the element and hence the Sobolev space in which it is embedded. I think what we need to do here is give the `WithMapping` constructor a `sobolev_space=L2` optional argument, and set `self._sobolev_space` to that and finally return that attribute here.

The alternative is to set the Sobolev space according to the mapping requested, but it's not obvious to me that that's safe.

Reply:
I think the way that sobolev spaces are used, `L2` is always safe? UFL uses it to determine if restricted elements are continuous across faces, not for determining which pullback to apply. So would agree that defaulting to `L2` is a good choice.

Reply:
I defaulted to L2, but in case WithMapping is a no-op we may return `self.wrapee.sobolev_space()`